File: code/py/src/leetcode/done/until20240301/p1422.py
class Solution:
    def maxScore(self, s: str) -> int:
        left_zeros, right_ones = [0], [0]
        for char in s[:-1]:
            left_zeros.append(1 + left_zeros[-1] if '0' == char else left_zeros[-1])
        for char in reversed(s[1:]):
            right_ones.insert(0, 1 + right_ones[0] if '1' == char else right_ones[0])
        del left_zeros[0], right_ones[-1]
        ans = 0
        for i in range(len(s) - 1):
            ans = max(ans, left_zeros[i] + right_ones[i])
        return ans


eg_s = '011101'
print(Solution().maxScore(s = eg_s))

File: code/kt/src/onlinejudge/leetcode/solved/P1422.kt
package onlinejudge.leetcode.solved

import kotlin.math.max

class P1422 {
    fun maxScore(s: String): Int {
        val leftZeros: IntArray = IntArray(s.length)
        for (p in 0 until s.lastIndex) {
            leftZeros[p + 1] = leftZeros[p] + if ('0' == s[p]) 1 else 0
        }
        val rightOnes: IntArray = IntArray(s.length)
        for (p in s.lastIndex downTo 1) {
            rightOnes[p - 1] = rightOnes[p] + if ('1' == s[p]) 1 else 0
        }
        var ans: Int = 0
        for (p in 0 until s.lastIndex) {
            ans = max(ans, leftZeros[p + 1] + rightOnes[p])
        }
        return ans
    }
}